### ISSUE TITLE
Group Normalization

### ISSUE THREAD

**Issue by bayedieng** [Tracel-AI/burn#685]
### Feature description

Group Normalization as described in https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html

### Feature motivation
Group Normalization is quite present in lots of the more recent SOTA deep learning models and performs quite well, would be a nice addition to have.
 
### Suggest A Solution

Here's a rough example from a project I'm working assuming a 4d tensor. Not sure how I'd make it with ```n``` dimensions agnostic though:

```rust
#[derive(Config)]
pub struct GroupNormConfig {
    pub num_groups: usize,
    
    pub num_channels: usize,
    
    #[config(default = 1e-5)]
    pub epsilon: f32,
    
    #[config(default = true)]
    pub affine: bool,
}

impl GroupNormConfig {
    pub fn init<B: Backend>(&self) -> GroupNorm<B> {
        match self.affine {
            true => {
                if self.num_channels % self.num_groups == 0 {
                    panic!("Groups and channels have to be divisible by each other")
                }
                
                let gamma = Tensor::ones([1, self.num_channels, 1, 1]);
                let beta = Tensor::zeros([1, self.num_channels, 1, 1]);
                 GroupNorm {
                    num_groups: self.num_groups,
                    num_channels: self.num_channels,
                    epsilon: self.epsilon,
                    affine: self.affine,
                    gamma: Some(Param::from(gamma)),
                    beta: Some(Param::from(beta))
                }
            },
            false => {
                if self.num_channels % self.num_groups == 0 {
                    panic!("Groups and channels have to be divisible by each other")
                }
                
                  GroupNorm {
                    num_groups: self.num_groups,
                    num_channels: self.num_channels,
                    epsilon: self.epsilon,
                    affine: self.affine,
                    gamma: None,
                    beta: None
                }
            }
        }


    }
}
#[derive(Debug, Module)]
pub struct GroupNorm<B: Backend> {
    num_groups: usize,
    num_channels: usize,
    epsilon: f32,
    affine: bool, 
    gamma: Option<Param<Tensor<B, 4>>>,
    beta: Option<Param<Tensor<B, 4>>>
}

impl<B: Backend>GroupNorm<B> {
    pub fn forward(&self, x: Tensor<B, 4>) -> Tensor<B, 4> {
        let x_shape = x.dims();
        let x = x.reshape([x_shape[0], self.num_groups, (self.num_channels / self.num_groups) as usize, x_shape[2], x_shape[3]]);
        
        let x_mean = x.clone()
        .mean_dim(2)
        .mean_dim(3)
        .mean_dim(4);
        
        let x_var = x.clone()
        .var(2)
        .var(3)
        .var(4);
        
        let x = (x - x_mean) / (x_var + self.epsilon).sqrt();
        let x = x.reshape(x_shape);
        
        match self.affine {
            true => {
                let x = self.gamma.clone().unwrap().val() * x + self.beta.clone().unwrap().val();
                x
            },
            false => x
        }
    }
}
```
